File: wrapper/src/descriptors.cc
#include "rust/cxx.h"
#include <GraphMol/Descriptors/Property.h>
#include <GraphMol/ROMol.h>


namespace RDKit {
    using Descriptors::Properties;

    std::shared_ptr<Properties> new_properties() {
        return std::shared_ptr<Properties>(new Properties());
    }

    std::unique_ptr<std::vector<std::string>> get_property_names(const std::shared_ptr<Properties> &props) {
        std::vector<std::string> names = props->getPropertyNames();
        std::vector<std::string> *names_heap = new std::vector<std::string>(names);
        return std::unique_ptr<std::vector<std::string>>(names_heap);
    }

    std::unique_ptr<std::vector<double>> compute_properties(const std::shared_ptr<Properties> &props, const std::shared_ptr<ROMol> &mol) {
        std::vector<double> computed = props->computeProperties(*mol);
        auto computed_heap = new std::vector<double>(computed);
        return std::unique_ptr<std::vector<double>>(computed_heap);
    }
}

File: src/bridge/descriptors.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    unsafe extern "C++" {
        include!("wrapper/include/ro_mol.h");
        include!("wrapper/include/descriptors.h");

        pub type ROMol = crate::ro_mol_ffi::ROMol;
        pub type Properties;

        pub fn new_properties() -> SharedPtr<Properties>;
        pub fn get_property_names(
            properties: &SharedPtr<Properties>,
        ) -> UniquePtr<CxxVector<CxxString>>;
        pub fn compute_properties(
            properties: &SharedPtr<Properties>,
            mol: &SharedPtr<ROMol>,
        ) -> UniquePtr<CxxVector<f64>>;
    }
}
